api/versions.go

```diff
@@ -1,129 +1,153 @@
 	"github.com/ONSdigital/go-ns/log"
 	"github.com/gorilla/mux"
 	"github.com/pkg/errors"
 )
 
+var (
+	publishedVersionCollectionIDInvalidErr  = errors.New("Unexpected collection_id in published version")
+	associatedVersionCollectionIDInvalidErr = errors.New("Missing collection_id for association between version and a collection")
+	versionStateInvalidErr                  = errors.New("Incorrect state, can be one of the following: edition-confirmed, associated or published")
+	versionPublishedAction                  = "versionPublished"
+	versionDownloadsGenerated               = "versionDownloadsGenerated"
+)
+
+type versionDetails struct {
+	datasetID string
+	edition   string
+	version   string
+}
+
+func (v versionDetails) baseAuditParams() common.Params {
+	return common.Params{"dataset_id": v.datasetID, "edition": v.edition, "version": v.version}
+}
+
+func (v versionDetails) baseLogData() log.Data {
+	return log.Data{"dataset_id": v.datasetID, "edition": v.edition, "version": v.version}
+}
+
 func (api *DatasetAPI) getVersions(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
 	vars := mux.Vars(r)
 	id := vars["id"]
 	editionID := vars["edition"]
 	logData := log.Data{"dataset_id": id, "edition": editionID}
 	auditParams := common.Params{"dataset_id": id, "edition": editionID}
 
-	if auditErr := api.auditor.Record(r.Context(), getVersionsAction, actionAttempted, auditParams); auditErr != nil {
-		handleAuditingFailure(w, auditErr, logData)
+	if auditErr := api.auditor.Record(r.Context(), getVersionsAction, audit.Attempted, auditParams); auditErr != nil {
+		audit.LogActionFailure(ctx, getVersionsAction, audit.Attempted, auditErr, logData)
+		handleVersionAPIErr(ctx, errs.ErrInternalServer, w, logData)
 		return
 	}
 
-	authorised, logData := api.authenticate(r, logData)
+	b, err := func() ([]byte, error) {
+		authorised, logData := api.authenticate(r, logData)
 
-	var state string
-	if !authorised {
-		state = models.PublishedState
-	}
-
-	if err := api.dataStore.Backend.CheckDatasetExists(id, state); err != nil {
-		log.ErrorC("failed to find dataset for list of versions", err, logData)
-		if auditErr := api.auditor.Record(r.Context(), getVersionsAction, actionUnsuccessful, auditParams); auditErr != nil {
-			handleAuditingFailure(w, auditErr, logData)
-			return
+		var state string
+		if !authorised {
+			state = models.PublishedState
 		}
-		handleErrorType(versionDocType, err, w)
-		return
-	}
 
-	if err := api.dataStore.Backend.CheckEditionExists(id, editionID, state); err != nil {
-		log.ErrorC("failed to find edition for list of versions", err, logData)
-		if auditErr := api.auditor.Record(r.Context(), getVersionsAction, actionUnsuccessful, auditParams); auditErr != nil {
-			handleAuditingFailure(w, auditErr, logData)
-			return
+		if err := api.dataStore.Backend.CheckDatasetExists(id, state); err != nil {
+			logError(ctx, errors.WithMessage(err, "failed to find dataset for list of versions"), logData)
```

Comment:
I did consider this however `log.ErrorCtx()` doesn't add the caller/user identity - so I have updated my `logInfo` & `logError` func to call the new functions after adding these fields.

Reply:
Ah sorry, so instead there are the following `LogError` and `LogInfo` functions in go-ns audit package which handle the caller/user identities.

So it would be `audit.LogError(ctx, ...)` and `audit.LogInfo(ctx, ...)`, see [here](https://github.com/ONSdigital/go-ns/blob/5cee3f2646232f0706d6ab2447ef617a51fc94cf/audit/log.go#L29)